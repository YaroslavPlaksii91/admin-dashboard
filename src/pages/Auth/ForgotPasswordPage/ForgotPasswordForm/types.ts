export type ForgotPasswordFormData = {
  email: string;
};

export type ForgotPasswordFormProps = {
  setIsSubmitted: (isSubmitted: boolean) => void;
};
