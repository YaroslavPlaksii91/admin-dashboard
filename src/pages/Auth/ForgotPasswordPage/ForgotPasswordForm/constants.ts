export const FORGOT_PASSWORD_FIELDS = Object.freeze({
  EMAIL: 'email',
});

export const FORGOT_PASSWORD_FIELDS_CONFIG = Object.freeze({
  [FORGOT_PASSWORD_FIELDS.EMAIL]: {
    label: 'Email',
    type: 'email',
    name: 'email',
    placeholder: 'Email address',
  },
});

export const FORGOT_PASSWORD_FORM_TEST_ID = 'forgotPasswordForm';
