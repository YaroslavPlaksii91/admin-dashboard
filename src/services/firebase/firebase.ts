import { initializeApp } from 'firebase/app';
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  sendPasswordResetEmail,
  confirmPasswordReset,
  getAuth,
} from 'firebase/auth';

const firebaseConfig = {
  apiKey: import.meta.env.VITE_FIREBASE_API_KEY,
  authDomain: import.meta.env.VITE_FIREBASE_AUTH_DOMAIN,
  projectId: import.meta.env.VITE_FIREBASE_PROJECT_ID,
  storageBucket: import.meta.env.VITE_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: import.meta.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
  appId: import.meta.env.VITE_FIREBASE_APP_ID,
  measurementId: import.meta.env.VITE_FIREBASE_MEASUREMENT_ID,
};

export const app = initializeApp(firebaseConfig);
export const auth = getAuth(app);

export const createUser = async (email: string, password: string) => {
  return createUserWithEmailAndPassword(auth, email, password);
};

export const signInUser = async (email: string, password: string) => {
  return signInWithEmailAndPassword(auth, email, password);
};

export const sendResetEmail = async (email: string) => {
  return sendPasswordResetEmail(auth, email);
};

export const confirmThePasswordReset = async (
  oobCode: string,
  newPassword: string,
) => {
  if (!oobCode && !newPassword) return;

  return confirmPasswordReset(auth, oobCode, newPassword);
};
