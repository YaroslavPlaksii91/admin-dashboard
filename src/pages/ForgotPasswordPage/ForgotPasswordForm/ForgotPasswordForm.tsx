import { FC } from 'react';
import { useNavigate } from 'react-router';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Button } from '@mui/material';

import { EMAIL_REGEX } from '@utils/constants';
import { ROUTES } from '@routes/constants';
import { getUser } from '@services/localeStorage';
import { FormInput } from '@components/FormInput/FormInput';

import { ForgotPasswordFormData, ForgotPasswordFormProps } from './types';
import {
  FORGOT_PASSWORD_FIELDS,
  FORGOT_PASSWORD_FIELDS_CONFIG,
} from './constants';

export const ForgotPasswordForm: FC<ForgotPasswordFormProps> = ({
  setIsSubmitted,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ForgotPasswordFormData>();
  const navigate = useNavigate();

  const onSubmit: SubmitHandler<ForgotPasswordFormData> = data => {
    const user = getUser(data.email);

    if (!user) {
      return console.error('User not found!');
    }

    setIsSubmitted(true);

    setTimeout(() => {
      navigate(ROUTES.RESET_PASSWORD_PAGE, { state: { from: data.email } });
    }, 3000);
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} noValidate>
      <FormInput
        {...FORGOT_PASSWORD_FIELDS_CONFIG[FORGOT_PASSWORD_FIELDS.EMAIL]}
        register={register('email', {
          required: 'Email is required',
          pattern: {
            value: EMAIL_REGEX,
            message: 'Invalid email address',
          },
        })}
        errors={errors}
      />

      <Button variant="contained" type="submit" fullWidth>
        Send
      </Button>
    </form>
  );
};
