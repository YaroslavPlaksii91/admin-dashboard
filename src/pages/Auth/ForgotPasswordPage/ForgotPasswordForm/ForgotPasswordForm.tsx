import { FC } from 'react';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Button } from '@mui/material';

import { EMAIL_REGEX } from '@utils/constants';
import { sendResetEmail } from '@services/firebase/firebase';
import { FormInput } from '@components/FormInput/FormInput';

import { ForgotPasswordFormData, ForgotPasswordFormProps } from './types';
import {
  FORGOT_PASSWORD_FIELDS,
  FORGOT_PASSWORD_FIELDS_CONFIG,
} from './constants';

export const ForgotPasswordForm: FC<ForgotPasswordFormProps> = ({
  setIsSubmitted,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ForgotPasswordFormData>();

  const onSubmit: SubmitHandler<ForgotPasswordFormData> = async data => {
    try {
      await sendResetEmail(data.email);

      setIsSubmitted(true);
    } catch (error) {
      console.error('Error: ', error);
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} noValidate>
      <FormInput
        {...FORGOT_PASSWORD_FIELDS_CONFIG[FORGOT_PASSWORD_FIELDS.EMAIL]}
        register={register('email', {
          required: 'Email is required',
          pattern: {
            value: EMAIL_REGEX,
            message: 'Invalid email address',
          },
        })}
        errors={errors}
      />

      <Button variant="contained" type="submit" fullWidth>
        Send
      </Button>
    </form>
  );
};
