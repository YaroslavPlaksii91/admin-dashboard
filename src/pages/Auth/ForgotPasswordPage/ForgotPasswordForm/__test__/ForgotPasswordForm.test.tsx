import {
  cleanup,
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';

import { sendResetEmail } from '@services/firebase/firebase';

import { ForgotPasswordForm } from '../ForgotPasswordForm';

jest.mock('@services/firebase/firebase', () => ({
  sendResetEmail: jest.fn(),
}));

describe('ForgotPasswordForm component', () => {
  afterEach(cleanup);

  const setIsSubmitted = jest.fn();

  it('should be rendered', () => {
    const { asFragment } = render(
      <ForgotPasswordForm setIsSubmitted={setIsSubmitted} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(<ForgotPasswordForm setIsSubmitted={setIsSubmitted} />);

    const emailInput = screen.getByLabelText('Email');
    const submitButton = screen.getByRole('button', { name: 'Send' });

    expect(emailInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });

  it('submits the form with valid email', async () => {
    render(<ForgotPasswordForm setIsSubmitted={setIsSubmitted} />);

    const emailInput = screen.getByLabelText('Email');
    const submitButton = screen.getByRole('button', { name: 'Send' });

    fireEvent.change(emailInput, {
      target: { value: 'example@mail.com' },
    });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(sendResetEmail).toHaveBeenCalledWith('example@mail.com');
    });
  });

  it('displays error message for invalid email', async () => {
    render(<ForgotPasswordForm setIsSubmitted={setIsSubmitted} />);

    const emailInput = screen.getByLabelText('Email');
    const submitButton = screen.getByRole('button', { name: 'Send' });

    fireEvent.change(emailInput, { target: { value: 'invalid-email' } });

    const errorMessage = screen.queryByText('Invalid email address');
    expect(errorMessage).not.toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(
      await screen.findByText('Invalid email address'),
    ).toBeInTheDocument();
  });

  it('will not be submitted if email is empty string', async () => {
    render(<ForgotPasswordForm setIsSubmitted={setIsSubmitted} />);

    const submitButton = screen.getByRole('button', { name: 'Send' });

    const errorMessage = screen.queryByText('Email is required');
    expect(errorMessage).not.toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(await screen.findByText('Email is required')).toBeInTheDocument();
  });
});
