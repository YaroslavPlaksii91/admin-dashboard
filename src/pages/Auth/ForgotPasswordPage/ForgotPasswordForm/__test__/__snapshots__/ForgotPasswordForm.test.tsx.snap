// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPasswordForm component should be rendered 1`] = `
<DocumentFragment>
  <form
    data-testid="forgotPasswordForm"
    novalidate=""
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
        id=":r0:-label"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="email"
          placeholder="Email address"
          type="email"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Send
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
